docs/api_design_v1beta2.md

```diff
@@ -0,0 +1,542 @@
+
+## Out of scope
+
+* The detailed implementation of new API is not discussed in this doc.
+
+## Roles:
+
+This document considers two user role:
+
+* **Cluster Administrator:** The cluster administrator (ops) is responsible for administration of entire clusters. They manage the lifecycle of AppMesh mesh and provide them to application developers.
+* **Application Developer**: The application developer (dev) is responsible for defining their application configuration and service composition.
+
+## Use cases:
+
+* As a Cluster Administrator for multi-tenant EKS cluster, I should be able to allocate a service mesh and control which users can create resources in this mesh.
+* As an Application Developer, I should be able to define the service routing logic and backend configuration for my application without Cluster Administrator's intervention.
+
+## Tenets:
+
+* The API should follow [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+* The API should be unambiguous, there should be only one obvious way for given task
+* The API should be explicit instead of rely on implicit logic
+* The API should be simple for 99% use case
+* The API should be open for future extensions
+
+## Design
+
+### API overview
+
+We’ll use four CRD to represent AppMesh objects:
+
+* Mesh: Represents an AppMesh mesh object.
+* VirtualNode: Represents an AppMesh virtualNode object.
+* VirtualService: Represents an AppMesh virtualService object, which can use either VirtualRouter or VirtualNode as provider.
+* VirtualRouter: Represents an AppMesh virtualRouter object and embeds virtualRoute.
+
+![Image: api_overview.png](api_overview.png)
+
+Detailed changes are discussed below:
+
+### Decouple k8s resource name from AppMesh resource name
+
+Resource names for k8s resources should be decoupled from appMesh resource names. Users should be able to freely denote the resource name for k8s resources independently from appMesh resource names. **Sensible default for appMesh resource names can be derived from k8s resource name**.
+
+1. Mesh resource have a field “**awsName**” to denote the AppMesh name for mesh. It defaults to be same as the name of mesh k8s resource.  
+
+    Note: we may consider use other field name like “meshName”, “awsMeshName”. we choose “awsName” over simply “name” to avoid confusing.
+
+    This enables use cases like use same resource name(e.g. “*global*”) for the k8s Mesh resource in different EKS clusters, while maps to different appMesh name.
+
+    ```yaml
+    apiVersion: appmesh.k8s.aws/v1beta2
+    kind: Mesh
+    metadata:
+      name: global
+    spec:
+      awsName: my-cluster-mesh
+    ```
+
+1. VirtualService resource have an array field “**hosts**” to denote the AppMesh names for VirtualService **For each host, we’ll transparently create an appMesh VirtualService object**. It defaults to have single entry of the VirtualService resource’s name.
```

Comment:
What's the motivation for this? I see below you said `we cannot provides the ability to expose a virtualService into multiple hostNames conveniently`, but what's the use case for that? Is it just the example where you're representing multiple parts of the fully qualified name?

There are implications around auth, processing the mesh, limits, UX, etc.

Reply:
Also interested in opinion from @bcelenza

Reply:
Without multiple hosts there is no way to register a service the way Kubernetes works e.g.:
```
    - myservice
    - myservice.default
    - myservice.default.svc
    - myservice.default.svc.cluster.local
```

Reply:
@stefanprodan 
Yeah, it's a tradeoff from k8s experience from AppMesh experience.
There is still a way to do it by create multiple VirtualService reference same virtualNode/virtualRoute. And I believe at least 90% case only use a single host name to reference a service.

Reply:
We use the main Virtual Service name (i.e. the resource name) to build the service graph in App Mesh, so each new Virtual Service is another potential backend in someone’s virtual node, another place to need a client policy, etc. But we’ve also discussed making name resolution work more like k8s everywhere, and with Envoy responding to DNS queries (https://github.com/envoyproxy/envoy/issues/6748 ) we could make this work anywhere.

@stefanprodan @M00nF1sh For multiple hosts in the k8s way, are they always some variant of the FQDN (e.g. myservice.default.svc.cluster.local, myservice.default), or could they be entirely separate FQDNs (e.g. myservice.default.svc.cluster.local, anotherservice.elsewhere.svc.cluster.local)? 

We could conceivably support the former in the Virtual Service through some additional API fields (e.g. a field for aliases, or just a boolean that enables all the variants @stefanprodan is describing).

Reply:
@bcelenza it can be "myservice" as well. The way it works is pod's /etc/resolve.conf is automatically configured with `default.svc.cluster.local`. so when pods making requests to "myservice", it also works, and resolves to the backend of `myservice.default.svc.cluster.local`
BTW, we have changed the virtualHost to just have a "awsName" field instead of hosts array. (we can add alias(or something else) array when appmesh supports)